### PR TITLE
update checkout views to support course entitlement products.

### PR DESCRIPTION
**Pull request by mjfrey**
Lookup course product meta-data from discovery for both seats as well as course entitlements.

Update checkout views to support both product types.

LEARNER-2655

### CODE REVIEW

File: ecommerce/courses/utils.py [hunk 1]
```diff
@@ -22,17 +24,25 @@ def mode_for_product(product):
...
     partner_short_code = site.siteconfiguration.partner.short_code
-    cache_key = 'courses_api_detail_{}{}'.format(course_key, partner_short_code)
+    cache_key = 'courses_api_detail_{}{}'.format(key, partner_short_code)
     cache_key = hashlib.md5(cache_key).hexdigest()
-    course_run = cache.get(cache_key)
-    if not course_run:  # pragma: no cover
-        course_run = api.course_runs(course_key).get(partner=partner_short_code)
-        cache.set(cache_key, course_run, settings.COURSES_API_CACHE_TIMEOUT)
-    return course_run
+    course = cache.get(cache_key)
```
Comment (staubina) [inline ecommerce/courses/utils.py]: Nit, Should we use a different name than course here, as we are talking about either a course_run or a Course.

**Review by staubina — APPROVED** [review]
I do not see anything in here that is concerning here.

**Comment by mjfrey** [comment]
@edx/helio please review.

**Review by McKenzieW — APPROVED** [review]
LGTM 👍

File: ecommerce/coupons/tests/mixins.py [hunk 2]
```diff
@@ -55,6 +55,31 @@ def mock_course_run_detail_endpoint(self, course_run, discovery_api_url, course_
...
+                "uuid": course.attr.UUID,
+                "title": course.title,
+                "short_description": 'Foo',
+                "image": {
+                    "src": "/path/to/image.jpg",
+                },
+            }
+
+        course_info_json = json.dumps(course_info)
+        course_url = '{}courses/{}/'.format(
```
Comment (bderusha) [inline ecommerce/coupons/tests/mixins.py:72]: Prefer names variables. `'{root_url}courses/{uuid}/'.format...`

File: ecommerce/courses/tests/test_utils.py [hunk 3]
```diff
@@ -43,22 +45,36 @@ def test_mode_for_product(self, certificate_type, id_verification_required, mode
...
+            self.mock_course_run_detail_endpoint(
+                course, discovery_api_url=self.site_configuration.discovery_api_url
+            )
+        else:
+            product = create_or_update_course_entitlement(
+                'verified', 100, self.partner, 'foo-bar', 'Foo Bar Entitlement')
+            key = product.attr.UUID
+            self.mock_course_detail_endpoint(product, discovery_api_url=self.site_configuration.discovery_api_url)
+
+        cache_key = 'courses_api_detail_{}{}'.format(key, self.site.siteconfiguration.partner.short_code)
```
Comment (bderusha) [inline ecommerce/courses/tests/test_utils.py:67]: Same here.  Prefer named variables.

File: ecommerce/courses/utils.py [hunk 4]
```diff
@@ -22,17 +24,25 @@ def mode_for_product(product):
...
+    """ Get course or course_run information from Discovery Service and cache """
+    if product.is_course_entitlement_product:
+        key = product.attr.UUID
+    else:
+        key = CourseKey.from_string(product.attr.course_key)
+
     api = site.siteconfiguration.discovery_api_client
     partner_short_code = site.siteconfiguration.partner.short_code
-    cache_key = 'courses_api_detail_{}{}'.format(course_key, partner_short_code)
+    cache_key = 'courses_api_detail_{}{}'.format(key, partner_short_code)
```
Comment (bderusha) [inline ecommerce/courses/utils.py:36]: prefer named variables.

**Review by bderusha — COMMENTED** [review]
Belated review:  One set of style guide updates that would be great to get baked into whatever your next ecom PR is.  Otherwise, nice work!

**Pull request merged**